raphtory-api/src/core/utils/time.rs

```diff
@@ -0,0 +1,218 @@
+    /// fields mandatory except for milliseconds and allows replacing the T with a space
+    fn try_into_time(self) -> Result<TimeIndexEntry, ParseTimeError> {
+        let rfc_result = DateTime::parse_from_rfc3339(self);
+        if let Ok(datetime) = rfc_result {
+            return Ok(TimeIndexEntry::from(datetime.timestamp_millis()));
+        }
+
+        let result = DateTime::parse_from_rfc2822(self);
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.timestamp_millis()));
+        }
+
+        let result = NaiveDate::parse_from_str(self, "%Y-%m-%d");
+        if let Ok(date) = result {
+            let timestamp = date
+                .and_hms_opt(00, 00, 00)
+                .unwrap()
+                .and_utc()
+                .timestamp_millis();
+            return Ok(TimeIndexEntry::from(timestamp));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%dT%H:%M:%S%.3f");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%dT%H:%M:%S%");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%d %H:%M:%S%.3f");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        let result = NaiveDateTime::parse_from_str(self, "%Y-%m-%d %H:%M:%S%");
+        if let Ok(datetime) = result {
+            return Ok(TimeIndexEntry::from(datetime.and_utc().timestamp_millis()));
+        }
+
+        Err(ParseTimeError::InvalidDateTimeString(self.to_string()))
+    }
+}
+
+pub trait TryIntoTimeNeedsSecondaryIndex: TryIntoTime {}
+
+impl TryIntoTimeNeedsSecondaryIndex for i64 {}
+
+impl<Tz: TimeZone> TryIntoTimeNeedsSecondaryIndex for DateTime<Tz> {}
+
+impl TryIntoTimeNeedsSecondaryIndex for NaiveDateTime {}
+
+impl TryIntoTimeNeedsSecondaryIndex for &str {}
+
+/// Used to handle automatic injection of secondary index if not explicitly provided
+pub enum InputTime {
+    Simple(i64),
+    Indexed(i64, usize),
+}
+
+impl InputTime {
+    pub fn set_index(self, index: usize) -> Self {
+        match self {
+            InputTime::Simple(time) => InputTime::Indexed(time, index),
+            InputTime::Indexed(time, _) => InputTime::Indexed(time, index),
+        }
+    }
+
+    pub fn as_time(&self) -> TimeIndexEntry {
+        match self {
+            InputTime::Simple(t) => TimeIndexEntry::new(*t, 0),
+            InputTime::Indexed(t, s) => TimeIndexEntry::new(*t, *s),
+        }
+    }
+}
+
+// Single time input refers to the i64 component of a TimeIndexEntry.
+// Essentially it is the timestamp without the secondary index of a TimeIndexEntry
```

Comment:
```suggestion
// Essentially it is the timestamp without the secondary index of a TimeIndexEntry.
```

What does 'essentially' mean here? It would be better to either say 'It is X' or to explain explicitly the nunances